### PR TITLE
[SPARK-32381][CORE][SQL][FOLLOWUP] More cleanup on HadoopFSUtils

### PR DESCRIPTION
**Pull request by sunchao**
### What changes were proposed in this pull request?

This PR is a follow-up of #29471 and does the following improvements for `HadoopFSUtils`:
1. Removes the extra `filterFun` from the listing API and combines it with the `filter`.
2. Removes `SerializableBlockLocation` and `SerializableFileStatus` given that `BlockLocation` and `FileStatus` are already serializable.
3. Hides the `isRootLevel` flag from the top-level API.

### Why are the changes needed?

Main purpose is to simplify the logic within `HadoopFSUtils` as well as cleanup the API.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing unit tests (e.g., `FileIndexSuite`)

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala
Comment (holdenk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala:219]: This seems to indicate the change needs some work.

**Review by holdenk — COMMENTED** [review]
Thanks for continuing to work on this, some initial questions.

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala
Comment (sunchao) [reply core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:64]: yup will do

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala
Comment (sunchao) [reply core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:172]: yeah a FS can choose not to implement it (although all the main ones override this). If not implemented it will fall back to the default impl in `FileSystem`, which basically calls `listStatus` and then `getFileBlockLocations` on each `FileStatus` received. The behavior is very similar to what this class is doing later on.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala
Comment (sunchao) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala:219]: Yes this test checks the case where a file was deleted after a `listStatus` call but before a subsequent `getFileBlockLocations` when locality info is needed. With the new impl, we'd call `listLocatedStatus` instead which will call `getFileBlockLocations` internally, and thus the `listLocatedStatus` call (as a whole) fails with `FileNotFoundException`. 

As explained in the PR description, the behavior will be different when `spark.sql.files.ignoreMissingFiles` is set, although I think we currently don't give any guarantee when there is missing files during listing, so either is acceptable? anyway, I'm happy to remove this change if there is any concern.

**Comment by sunchao** [comment]
Thanks @holdenk for the review. Yes this PR still needs a bit more work. Will update.

**Comment by viirya** [comment]
Jenkins, add to whitelist

**Comment by sunchao** [comment]
retest this please

**Comment by sunchao** [comment]
Removed the changes on file listing (may deal with it in a separate PR). @holdenk mind take another look? Thanks.

**Comment by sunchao** [comment]
Removed the file listing changes (might open another PR just for that). @holdenk mind take another look? thanks.

**Comment by steveloughran** [comment]
One thing to know here is that while listStatus etc is slow against object stores, it's not enough of a bandwidth killer than you get much speedup executing it across a spark cluster. Same process-many-threads: yes. Remote exec: no.

I am going to make this class: org.apache.hadoop.mapred.LocatedFileStatusFetcher @Public/Evolving: in this PR: https://github.com/apache/hadoop/pull/2324
 it does do multithread scanning today and is fairly stable. 

It's not the perfect API -I'd have a builder and actually return a remoteIterator of statuses which would block while new listing entries arrived, but it's there. With that PR will also collect aggregate stats on IO operations from any FS whose list calls provide them.

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala
Comment (steveloughran) [reply core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala]: HDFS and S3A both do this; ABFS merits minor optimisation too. Because they return a remote iterator they can do paged fetch of data
* HDFS/webHDFS: paged download for better scalability
* S3A (3.3.1+): async prefetch of next page of data

ABFS should copy the S3A approach; it's listing API is paged too. 

Better to rely on the FS to do the work *but make clear you expect the maintainers to do so*

Comment (sunchao) [reply core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala]: Thanks @steveloughran , yes I also think it's better to rely on the FileSystem-specific `listLocatedStatus` impl rather than having the logic here. However, the change seems to break a few assumptions in the test cases so I'll isolate it into another PR.

File: common/network-yarn/pom.xml
Comment (dongjoon-hyun) [inline common/network-yarn/pom.xml:77]: This looks like a leftover. Shall we clean up this?

File: core/pom.xml
Comment (dongjoon-hyun) [inline core/pom.xml:566]: indentation?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:69]: ?

File: external/kafka-0-10-token-provider/pom.xml
Comment (dongjoon-hyun) [inline external/kafka-0-10-token-provider/pom.xml:28]: ?

File: common/network-yarn/pom.xml
Comment (sunchao) [reply common/network-yarn/pom.xml:77]: Please ignore this - I'll revert this commit after the testing is done

File: external/kinesis-asl-assembly/pom.xml
Comment (dongjoon-hyun) [inline external/kinesis-asl-assembly/pom.xml:135]: ?

**Comment by sunchao** [comment]
@dongjoon-hyun sorry I was reusing this PR for some testing. It is restored now.

**Comment by sunchao** [comment]
@holdenk @dongjoon-hyun could you take another look at this? I think the test failure is unrelated.

**Review by holdenk — APPROVED** [review]
LGTM. If there are no other outstanding concerns from folks I'll try and merge next week.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by sunchao** [comment]
Thanks @holdenk for taking a look. Can you help to merge this when you have time?

**Comment by holdenk** [comment]
I meant to merge this awhile ago, my bad.
Let's make sure everything still passes in CI.
Jenkins retest this please.

**Comment by holdenk** [comment]
Jenkins retest this please.

**Comment by holdenk** [comment]
Can you rebase this?

**Comment by sunchao** [comment]
@holdenk sure - it's done.

**Comment by holdenk** [comment]
K8s failures are unrelated, this does not change any of the decommissioning logic. I'll work on a follow up to the decommissioning failures.

**Comment by sunchao** [comment]
Thanks @holdenk for the review & merge!

**Comment by dongjoon-hyun** [comment]
Hi, @holdenk and @sunchao .

Could you check Hadoop 2.7 failure?
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-sbt-hadoop-2.7-hive-2.3/1609/

**Comment by sunchao** [comment]
thanks @dongjoon-hyun , let me take a look.

**Comment by sunchao** [comment]
found potential issue and opened #30447

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:212]: @sunchao the `dirs` here may contain hidden directories. We still need to filter them before listing leaf files.

Comment (sunchao) [reply core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:212]: @gengliangwang you're right. Thanks for catching this! and sorry for introducing this regression.

**Pull request closed without merging**
